File: app/src/incomes/incomes.service.ts
import { Injectable } from '@nestjs/common';
import { InjectRepository } from '@nestjs/typeorm';
import { IIncome } from 'src/interfaces/common';
import log, { LogLevels } from 'src/logger';
import { Repository } from 'typeorm';

import IncomeEntity from './incomes.entity';

@Injectable()
export class IncomesService {
  constructor(
    @InjectRepository(IncomeEntity)
    private incomeRepository: Repository<IncomeEntity>,
  ) {}

  async getIncomesByPeriod(
    period: string,
    userId: number,
    accountID: number,
  ): Promise<{
    incomes: IncomeEntity[] | null;
    incomeGraphData: any;
    msg: string;
  }> {
    try {
      const incomes = await this.incomeRepository.findBy({
        period,
        user_id: userId,
        account_id: accountID,
      });
      const graphData = this.addGraphData(incomes);
      return {
        incomes: this.normalizeIncomes(incomes),
        incomeGraphData: graphData,
        msg: '',
      };
    } catch (err) {
      console.log(err);
      log(`From income service: ${err}`, LogLevels.ERROR);
      return { incomes: null, incomeGraphData: null, msg: err };
    }
  }

  async addIncome(
    income: IIncome,
    userId: number,
  ): Promise<{
    status: number;
    data: { income: IncomeEntity | null; msg: string };
  }> {
    const newIncome = new IncomeEntity();
    newIncome.title = income.title;
    newIncome.amount = income.amount;
    newIncome.budget_id = income.budget_id;
    newIncome.date = new Date(income.date);
    newIncome.description = income.description;
    newIncome.month = income.month;
    newIncome.source_id = income.source_id;
    newIncome.user_id = userId;
    newIncome.year = income.year;
    newIncome.account_id = income.account_id;
    newIncome.period =
      income.year + '-' + String(income.month).padStart(2, '0');
    try {
      await this.incomeRepository.save(newIncome);
      return { status: 202, data: { income: newIncome, msg: '' } };
    } catch (err) {
      console.log(err);
      log(`From income service: ${err}`, LogLevels.ERROR);
      return { status: 500, data: { income: null, msg: err } };
    }
  }

  async deleteIncome(incomeId: number): Promise<{
    status: number;
    data: { isDeleted: boolean; msg: string };
  }> {
    try {
      await this.incomeRepository.delete(incomeId);
      return { status: 200, data: { isDeleted: true, msg: '' } };
    } catch (err) {
      console.log(err);
      log(`From income service: ${err}`, LogLevels.ERROR);
      return { status: 500, data: { isDeleted: false, msg: err } };
    }
  }

  // TODO In develop
  async deleteIncomesBySourceID(source_id: number): Promise<{
    status: number;
    data: { isDeleted: boolean; msg: string };
  }> {
    try {
      await this.incomeRepository
        .createQueryBuilder()
        .delete()
        .from('income')
        .where('source_id=:source_id', { source_id })
        .execute();
      return { status: 200, data: { isDeleted: true, msg: '' } };
    } catch (err) {
      console.log(err);
      log(`From cost service: ${err}`, LogLevels.ERROR);
      return { status: 500, data: { isDeleted: false, msg: err } };
    }
  }

  addGraphData(incomes) {
    const items = incomes.sort((a, b) => {
      if (new Date(a.date) > new Date(b.date)) return -1;
      if (new Date(a.date) < new Date(b.date)) return 1;
      return 0;
    });
    const graphIncomes = [];
    const graphDays = [];
    const days = new Set();
    for (const cost of items) {
      days.add(new Date(cost.date).getUTCDate());
    }
    for (const day of days) {
      let sum = 0;
      for (const cost of items) {
        const day2 = new Date(cost.date).getUTCDate();
        if (day === day2) sum += cost.amount;
      }
      graphIncomes.push(sum);
      graphDays.push(day);
    }
    graphIncomes.reverse();
    graphDays.reverse();
    return {
      days: graphDays,
      items: graphIncomes,
    };
  }

  normalizeIncomes(incomes) {
    if (incomes.length > 0) {
      const items = [];
      const periods = new Set();
      incomes.map((item) => {
        periods.add(item.date.toString());
      });
      let gainByPeriod = 0;

      for (const period of periods) {
        const item = incomes.filter((item) => item.date.toString() === period);
        let gainByDay = 0;
        for (const i of item) {
          gainByDay += i.amount;
          gainByPeriod += i.amount;
        }
        items.push({ period, items: item, gainByDay, gainByPeriod });
      }
      return items.sort((a, b) => {
        if (new Date(a.period) > new Date(b.period)) return -1;
        if (new Date(a.period) < new Date(b.period)) return 1;
        return 0;
      });
    } else {
      return incomes;
    }
  }
}

File: analytics/services/incomes_service.py
from models.incomes_model import Incomes
from models.income_sources_model import IncomeSources
from services.db_service import DBService


class IncomesService:
    @staticmethod
    def get_incomes(user_id: int, period: str, account_id: int) -> dict:
        return IncomesService.normalize_incomes(DBService.get_incomes_by_period(user_id, period, account_id))


    @staticmethod
    def get_incomes_sources(user_id: int, account_id: int) -> list:
        return IncomesService.normalize_incomes_sources(DBService.get_incomes_sources_by_period(user_id, account_id))

    @staticmethod
    def normalize_incomes(incomes: list) -> dict:
        items = []
        periods = set(income.date.isoformat() for income in incomes)
        periods = sorted(periods, reverse=True)

        for period in periods:
            gainByDay = 0
            incomesInDate = []
            for income in incomes:
                if income.date.isoformat() == period:
                    incomesInDate.append(IncomesService.get_public_income(income))
                    gainByDay += income.amount

            items.append({
                'period': period,
                'items': incomesInDate,
                'gainByDay': gainByDay
            })
        graph_data = IncomesService.add_graph_data(incomes)
        statistics = IncomesService.calculate_month_sum(items)
        return {'incomes': items, 'graph_data': graph_data, 'statistics': statistics}

    @staticmethod
    def get_public_income(income: Incomes) -> dict:
        return {
            'amount': income.amount,
            'id': income.id,
            'source_id': income.source_id,
            'account_id': income.account_id,
            'title': income.title,
            'year': income.year,
            'month': income.month,
            'period': income.period,
            'description': income.description,
            'date': income.date,
            'budget_id': income.budget_id
        }

    @staticmethod
    def normalize_incomes_sources(incomes_sources: list) -> list:
        return [IncomesService.get_public_income_source(income_source) for income_source in incomes_sources]

    @staticmethod
    def get_public_income_source(income_source: IncomeSources) -> dict:
        return {
            'id': income_source.id,
            'account_id': income_source.account_id,
            'title': income_source.title,
            'description': income_source.description,
            'order': income_source.order
        }

    @staticmethod
    def add_graph_data(incomes: list):
        graph_amount = []
        graph_days = []
        days = set([income.date.isoformat() for income in incomes])

        for day in days:
            sum_of_costs = 0
            for income in incomes:
                day2 = income.date.isoformat()
                if day == day2:
                   sum_of_costs += income.amount
            graph_amount.append(sum_of_costs)
            graph_days.append(day)

        graph_days.reverse()
        graph_amount.reverse()
        return {
            'days': graph_days,
            'items': graph_amount,
        }

    @staticmethod
    def calculate_month_sum(incomes):
        if not incomes:
            return {'spentByMonth': 0}
        sum_of_month = 0
        for income in incomes:
            sum_of_month += income['gainByDay']

        return {'gainByMonth': sum_of_month}
